app/models/ont/run.rb

```diff
@@ -76,30 +87,72 @@ def cancel
       update(deactivated_at: DateTime.current)
     end
 
+    # Sets flowcells from an array of attributes (hash)
+    # If there is no id in attributes, a new flowcell will be created.
+    # If there is an id, existing flowcell will be updated.
+    # If there is no attributes hash for an existing flowcell, it will be
+    # deleted.
+    # This method is called before validations, therefore it must properly
+    # set up all flowcells to be saved together with the run. It must first
+    # do the deletions and then "build" flowcells and "assign" attributes
+    # without saving.
+
     def flowcell_attributes=(flowcell_options)
-      self.flowcells = flowcell_options.map do |attributes|
+      transform_flowcell_attributes(flowcell_options)
+
+      # Delete flowcells if attributes are not given
+      options_ids = flowcell_options.pluck(:id).compact
+      flowcells.each do |flowcell|
+        flowcells.delete(flowcell) if options_ids.exclude? flowcell.id
+      end
+
+      # Update existing flowcells or build new ones
+      flowcell_options.map do |attributes|
         if attributes[:id]
           update_flowcell(attributes)
         else
-          Ont::Flowcell.new(attributes)
+          create_flowcell(attributes)
         end
       end
     end
 
     private
 
+    # Change attributes before using them
+    # Strip and upcase flowcell_ids to accept case-insensitive barcodes.
+    def transform_flowcell_attributes(flowcell_options)
+      flowcell_options.each do |attributes|
+        # because it is a frozen string
+        if attributes[:flowcell_id]
```

Comment:
This method is used by flowcell_attributes= method of run. The run is responsible for assignment of them as they are nested attributes of the run (autosave is enabled for flowcells; we create or update the run to persist flowcells together). In this particular case, it is sanitising the incoming data, converting flowcell_id (barcode-like) to capitals because we allowed case-insensitive input.

flowcell_attributes= is called before validations, we need to modify flowcells so the run and its flowcells are validated together. I will talk about it in the other comment as well.